### ISSUE TITLE
[Sharp IM-DR420]: Renaming tracks does not work (works with netmdcli)

### ISSUE THREAD

**Issue by andyboeh** [cybercase/netmd-js#8]
Unfortunately, I cannot rename tracks using the Sharp IM-DR420. This works properly using vuori's linux-minidisc fork (`netmdcli rename`), so I suspect something different than in cybercase/webminidisc#12. How can I debug the webminidisc application to nail this down (I only use the version hosted by you)?

Thanks

**Comment by cybercase** [comment]
Hi @andyboeh 

this issue is likely related to netmd-js library so I moved it here.
I’m not sure how confident you are with the JS stack so I’ll leave you a few instructions to start debugging the library.

Keep in mind that you’ll need node 12 installed and working properly.

```bash
# checkout the repository
git clone git@github.com:cybercase/netmd-js.git
cd netmd-js/

npm i

npm run dev # this will start the typescript compiler in watch mode
```

from another terminal, in the same directory

```bash
# list the tracks
node dist/cli.js ls 

# rename the first track to "First Track"
node ./dist/cli.js rename 0 "First Track"
```

**Comment by andyboeh** [comment]
Thanks for the quick reply! This helps a lot and I can start working on it. Seems like it's going to be more work than expected: The Sharp seems to be pickier about the protocol than the Sony and it responds with `0x08` half of the time. I'll try to sort this out and let you know if I succeed. This also applies to the problem with downloading a track.

**Comment by andyboeh** [comment]
So, there's quite some difficulties with the Sharp device, and I hope you can shed some light on this:

In `netmd-interface.ts` in setupDownload https://github.com/cybercase/netmd-js/blob/cf7353baaf954ec8985a8fe5bea5e3ece3ae8a84/src/netmd-interface.ts#L626 the encrypted packet created this way contains 40 bytes of ciphertext. However, it should only contain 32 bytes, otherwise the Sharp device bails out.
You have a comment somewhere else reading "CAVEAT: don't know why, but python Crypto returns 8 bytes less that js Crypto". 

In order to transfer a track, I just cut off the remaining 8 bytes in setupDownload, but then the recording only contains static noise (which was to be expected). 

Do you have any idea what to do here?

**Comment by cybercase** [comment]
Unfortunately I haven’t dug much into the differences between crypto-js and the python Crypto library.

To port and debug this part of the code from the linux-minidisc project I’ve added logs at every step in the code to make sure that the output of my code matched the original.

I used the original Python code as a reference, instead of the C counterpart, so the part you should look at in the original code is the `setupDownload` function in `libnetmd.py` .
However, if you’re looking at the C version, the function you need to look at should be `netmd_secure_setup_download`, in `secure.c`

Maybe it could be useful to see if there’s any difference between the original and the ported function by calling them with the same arguments and by checking the output packet.

**Comment by andyboeh** [comment]
Thanks, did that and it was actually easy: Crypto-JS, by default, applies padding while PyCrypto doesn't. I disabled padding just for transferring the key and that's.
I heard static because I uploaded an invalid file ;)
I still have to fix a few other things (it crashes after uploading the file), then I'll create a PR for review.

**Comment by cybercase** [comment]
I assume this has been fixed in #9
